posts/2018-07-12-servant-dsl-typelevel.md

```diff
@@ -0,0 +1,664 @@
+go through the captures and add an argument for each of them. Request bodies,
+query parameters, headers? We could probably track them too, in a similar way.
+Or we could unify it all by basically building up and tracking actual servant
+API types through a `GADT` version of Endpoint's type argument, and do some of
+what servant does at the type-level, with everything else at the value-level.
+
+However, all those approaches have a big problem. Once you've made a decision,
+it is set in stone, in a way. To be more precise, with all these approaches,
+if you want to extend your web app description language you need to add a new
+constructor to your `Endpoint` type, and you then need to handle this new
+constructor in all you functions that pattern match on `Endpoint` constructors
+since they've instantly become partial. Not to mention that just the act of
+adding a constructor requires rebuilding the entire library. You cannot explore
+two different directions simultaneously without breaking code, you cannot add
+new constructs you hadn't thought of without touching the library, e.g just
+locally in a project of yours. Extensibility and modularity
+were central requirements as we had been bitten by the lack of them
+in libraries that we were using at the time.
+
+So... how do people build extensible/modular DSLs in Haskell? The next section
+talks about the general problem behind this and a solution that I read about
+that gets us halfway to servant.
+
+# The Expression Problem
+
+To quote [Phil Wadler](http://homepages.inf.ed.ac.uk/wadler/papers/expression/expression.txt):
+"_the expression problem is a new name for an old problem. The goal is to define
+a datatype by cases, where one can add new cases to the datatype and new
+functions over the datatype, without recompiling existing code, and while
+retaining static type safety (e.g., no casts)_".
+
+In Haskell, the standard approach to representing some domain is
+to define an algebraic data type for it. For a simple type of expressions
+with additions and integers, we usually do:
+
+``` haskell
+data Expr = I Integer | Add Expr Expr
+```
+
+and proceed to write what we call "interpreters", which in this case are just
+functions that take expressions as input and do something interesting with them.
+
+``` haskell
+eval :: Expr -> Integer
+eval (I n)     = n
+eval (Add a b) = eval a + eval b
+
+prettyPrint :: Expr -> String
+prettyPrint (I n) = show n
+prettyPrint (Add a b) = unwords [prettyPrint a, "+", prettyPrint b]
+```
+
+So, given an expression type, we can easily "add new functions over the
+data type", to reuse Phil's wording. We just write a new function.
+However, when the time comes to "add new cases to the data type", this approach
+becomes painful. A "new case" here means a new constructor for our `Expr` data
+type. Let's say we want to support multiplications too:
+
+``` haskell
+data Expr = I Integer | Add Expr Expr | Mul Expr Expr
```

Comment:
I'm tempted to complicate the example with:

```haskell
data Expr r where
   I :: Integer -> Expr Integer
   Add :: Expr Integer -> Expr Integer -> Expr Integer
  Equals :: Expr Integer -> Expr Integer -> Expr Bool 
```

Because uni-type DSLs are rare.

The rest should come out still nicely, with `TypeFamilies` as we do in servant (`FunctionalDependencies` would fit here more nice).

Reply:
I feel like this would complicate things for very little gain, at least as far as the post is concerned. The main point of that section is to show that this is not extensible, and I feel like switching to a GADT for no gain for an expression type we don't care about for more than 5 minutes doesn't accomplish anything in that direction.

Reply:
Maybe as an intermediate section? You don't discuss / show GADTs at all, yet you mention them here and there.

Reply:
Yes, because they break the extensibility, and I cover that quite early on. I'm afraid that if I show equivalent GADT implementations of things, it would make things too long without making progress on the main goal of the post. Would you minf if I leave this bit of feedback unaddressed?